### DISCUSSION TITLE
函数、工具调用不成功

### DISCUSSION THREAD

**Discussion by DingGao1124** [zai-org/ChatGLM3#1178] [Q&A]
![p1](https://github.com/THUDM/ChatGLM3/assets/136151131/1acf1bcc-9d4e-4d19-a12f-c03faee5cead)
![805bc3387e288fffeee26012dac9ffe](https://github.com/THUDM/ChatGLM3/assets/136151131/1c7239e1-f6f5-4f89-8f32-3037de5816a5)
![p2](https://github.com/THUDM/ChatGLM3/assets/136151131/3036614b-2187-442b-bcb7-9d4c25336575)

运行openai_demo中的api_server代码后，函数回调不成功，好像没有解析成功我绑定的函数，返回的stop_reason是“stop”而不是"function_call“，请问如何解决 ？

**Comment by zRzRzRzRzRzRzR** [comment]
你都没有传入tools当然没有了 把 你注释的部分换回来 用tools

**Reply by DingGao1124** [reply]
刚开始是用的tools，发现没效果，所以换成了function试试看（因为一月份的时候也跑过api_server.py，当时代码是支持函数调用的），还是没效果, 原先工具调用所用代码如下：
<img width="937" alt="image" src="https://github.com/THUDM/ChatGLM3/assets/136151131/247b09c9-770b-4334-bfb1-bb2bf60a593a">
以下是通过gpt3.5函数调用返回的正确结果图示，可以看到返回的是调用工具所需要的参数：
<img width="992" alt="image" src="https://github.com/THUDM/ChatGLM3/assets/136151131/6e904f57-3a86-418b-b0a1-73dafc686288">
但是如果把代理地址换成服务器上部署的chatglm3，就会返回如下图所示的内容，发现并没有触发工具调用：
<img width="995" alt="image" src="https://github.com/THUDM/ChatGLM3/assets/136151131/f8bfe17a-4855-499a-b4be-5f65134923d9">
而且后端fastapi打印的日志显示“工具调用解析失败”：
<img width="1001" alt="image" src="https://github.com/THUDM/ChatGLM3/assets/136151131/1713590e-2ce5-4076-88ee-4b34aea2375d">
